### ISSUE TITLE
Wrong diagnostic position in the narrowed buffer

### ISSUE THREAD

**Issue by muffinmad** [joaotavora/eglot#479]
Python file:

```python
from datetime import datetime


print()
```

Note eglot-warning on line 1.

Select `print()` and do `C-x n n` (`narrow-to-region`).

Cancel selection and make an edit, e.g. add and remove space at the end of the line.

Note eglot-warning "datetime imported but unused" on the line `print()`.

Not sure if this is eglot or flymake bug though.

**Comment by nemethf** [comment]
Does the following patch fix the issue for you?  If it does, maybe we can start to think about how to create a better fix that calls save-restriction and widen less frequently, for example, just once for every received LSP message.  Or maybe the overhead is negligible, I don't know.

```patch
diff --git a/eglot.el b/eglot.el
index 8a1d162..ccd7f94 100644
--- a/eglot.el
+++ b/eglot.el
@@ -1118,6 +1118,7 @@ be set to `eglot-move-to-lsp-abiding-column' (the default), and
   "Convert LSP position POS-PLIST to Emacs point.
 If optional MARKER, return a marker instead"
   (save-excursion
+    (save-restriction (widen)
     (goto-char (point-min))
     (forward-line (min most-positive-fixnum
                        (plist-get pos-plist :line)))
@@ -1130,7 +1131,7 @@ If optional MARKER, return a marker instead"
            col)
           (setq col 0))
         (funcall eglot-move-to-column-function col)))
-    (if marker (copy-marker (point-marker)) (point))))
+    (if marker (copy-marker (point-marker)) (point)))))
 
 (defun eglot--path-to-uri (path)
   "URIfy PATH."
```

**Comment by joaotavora** [comment]
@nemethf that;s 99% the right fix.  I don't have any reason to believe that `save-restriction` is expensive.  A naive conjecture of how it works tells me that it is just saving and restoring two integers.

**Comment by joaotavora** [comment]
Earlier I meant to say "99% sure".  It does solve @muffinmad's problem, but  indeed something is hiding in that 1%, because as soon as you add the space _another_ "unused import" gets added to the invisible first line. I think in this case it's Flymake failing to remove diagnostics out of scope.

**Comment by joaotavora** [comment]
> I think in this case it's Flymake failing to remove diagnostics out of scope.

I meant to say "outside the region".  And it wasn't Flymake's fault, it was the fault of a somewhat hackish solution to fix #159.  Anyway, should be fixed now.
